src/main/java/com/minecolonies/coremod/generation/defaults/DefaultConcreteMixerCraftingProvider.java

```diff
@@ -0,0 +1,64 @@
+package com.minecolonies.coremod.generation.defaults;
+
+import com.minecolonies.api.colony.jobs.ModJobs;
+import com.minecolonies.api.crafting.ItemStorage;
+import com.minecolonies.coremod.generation.CustomRecipeProvider;
+import net.minecraft.data.DataGenerator;
+import net.minecraft.data.IFinishedRecipe;
+import net.minecraft.item.DyeColor;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.Items;
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.registries.ForgeRegistries;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.Consumer;
+
+/** Datagen for concrete mixer crafterrecipes */
+public class DefaultConcreteMixerCraftingProvider extends CustomRecipeProvider
+{
+    public DefaultConcreteMixerCraftingProvider(@NotNull final DataGenerator generatorIn)
+    {
+        super(generatorIn);
+    }
+
+    @Override
+    protected void registerRecipes(@NotNull final Consumer<IFinishedRecipe> consumer)
+    {
+        final List<ItemStorage> input = new ArrayList<>();
+        input.add(new ItemStorage(new ItemStack(Items.SAND, 4)));
+        input.add(new ItemStorage(new ItemStack(Items.GRAVEL, 4)));
+
+        for (final DyeColor color : DyeColor.values())
+        {
+            final String prefix = color.name().toLowerCase(Locale.ROOT);
+            final Item powder = ForgeRegistries.ITEMS.getValue(new ResourceLocation(prefix + "_concrete_powder"));
+            final Item concrete = ForgeRegistries.ITEMS.getValue(new ResourceLocation(prefix + "_concrete"));
+            final Item dye = ForgeRegistries.ITEMS.getValue(new ResourceLocation(prefix + "_dye"));
+
+            if (powder == null || concrete == null || dye == null)
+            {
+                throw new IllegalStateException("Missing items for " + color.getString());
+            }
+
+            final List<ItemStorage> customInput = new ArrayList<>(input);
+            customInput.add(new ItemStorage(new ItemStack(dye)));
+
+            CustomRecipeBuilder.create(ModJobs.CONCRETE_ID.getPath(), powder.getRegistryName().getPath())
+                    .inputs(customInput)
+                    .result(new ItemStack(powder, 8))
+                    .build(consumer);
+
+            CustomRecipeBuilder.create(ModJobs.CONCRETE_ID.getPath(), concrete.getRegistryName().getPath())
+                    .inputs(Collections.singletonList(new ItemStorage(new ItemStack(powder))))
+                    .result(new ItemStack(concrete))
+                    //.intermediate(Blocks.WATER)

```

Comment:
FYI, wasn't sure if this should be included or not.  It makes a certain amount of logical sense for it to be there, but the previously existing recipes did not set it.  And currently neither the RS crafting logic nor the JEI rendering support it.

Reply:
yeah, this is all special case handling. One of these days, as we modularize the crafting code, we should put in a intermediate handler, and allow a broader set of intermediates besides just air and furnace.

Reply:
Thinking about it, it might make sense to include it in the future, and update the crafter to trigger off the water for the special case handling, rather than use the specific items.

Reply:
Yeah, that's why I left it commented out, as a reminder for the future.

Reply:
Add a TODO to it then.